pipelines/tekton/pipeplines_provider.go

```diff
@@ -0,0 +1,210 @@
+// After the PipelineRun is being intitialized, the progress of the PipelineRun is being watched and printed to the output.
+func (pp *PipelinesProvider) Run(ctx context.Context, f fn.Function) error {
+	pp.progressListener.Increment("Creating Pipeline resources")
+
+	client, err := NewTektonClient()
+	if err != nil {
+		return err
+	}
+
+	err = k8s.CreatePersistentVolumeClaim(ctx, getPipelinePvcName(f), pp.namespace, corev1.ReadWriteOnce, *resource.NewQuantity(500, resource.DecimalSI))
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("problem creating persistent volume claim: %v", err)
+		}
+	}
+
+	_, err = client.Pipelines(pp.namespace).Create(ctx, generatePipeline(f), metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			if errors.IsNotFound(err) {
+				return fmt.Errorf("problem creating pipeline, missing tekton?: %v", err)
+			}
+			return fmt.Errorf("problem creating pipeline: %v", err)
+		}
+	}
+
+	_, err = k8s.GetSecret(ctx, getPipelineSecretName(f), pp.namespace)
+	if errors.IsNotFound(err) {
+		pp.progressListener.Stopping()
+		creds, err := pp.containerRegistryCredentialsCallback()
+		if err != nil {
+			return err
+		}
+		pp.progressListener.Increment("Creating Pipeline resources")
+
+		err = k8s.CreateDockerRegistrySecret(ctx, getPipelineSecretName(f), pp.namespace, creds.Username, creds.Password, creds.Server)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return fmt.Errorf("problem in creating secret: %v", err)
+	}
+
+	err = k8s.CreateServiceAccountWithSecret(ctx, getPipelineBuilderServiceAccountName(f), pp.namespace, getPipelineSecretName(f))
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("problem in creating service account: %v", err)
+		}
+	}
+
+	err = k8s.CreateClusterRoleBindingForServiceAccount(ctx, getPipelineDeployerClusterRoleBindingName(f, pp.namespace), pp.namespace, getPipelineBuilderServiceAccountName(f), "knative-serving-namespaced-edit")
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("problem in creating cluster role biding: %v", err)
+		}
+	}
+
+	pp.progressListener.Increment("Running Pipeline with the Function")
+	pr, err := client.PipelineRuns(pp.namespace).Create(ctx, generatePipelineRun(f), metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("problem in creating pipeline run: %v", err)
+	}
+
+	err = pp.watchPipelineRunProgress(pr)
+	if err != nil {
+		return fmt.Errorf("problem in watching started pipeline run: %v", err)
+	}
+
+	pr, err = client.PipelineRuns(pp.namespace).Get(ctx, pr.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("problem in retriving pipeline run status: %v", err)
+	}
+
+	if pr.Status.GetCondition(apis.ConditionSucceeded).Status == corev1.ConditionFalse {
+		return fmt.Errorf("function pipeline run has failed, please inspect logs of Tekton PipelineRun \"%s\"", pr.Name)
+	} else {
```

Comment:
minor nit: the contents of this else statement can be moved out (the if statement returns)